### ISSUE TITLE
Single GPU v.s. Multiple GPUs: Results not compatible

### ISSUE THREAD

**Issue by Palang2014** [dmlc/mxnet#1338]
I have run a simple experiment on CIFAR-10 using 1 GPU and 4GPUs. The time per epoch drops from 37 sec to 12 sec, which is good. But the Train and Validation accuracy are not the same. Any idea why ?!

E.g., for my toy example:

1 GPU: 
```python
2016-01-22 14:22:03,897 Node[0] Epoch[4] Train-accuracy=0.449129
2016-01-22 14:22:03,897 Node[0] Epoch[4] Time cost=37.519
2016-01-22 14:22:05,303 Node[0] Epoch[4] Validation-accuracy=0.467147
```

4 GPUs:
```python
2016-01-22 14:44:06,682 Node[0] Epoch[4] Train-accuracy=0.396080
2016-01-22 14:44:06,682 Node[0] Epoch[4] Time cost=12.148
2016-01-22 14:44:07,227 Node[0] Epoch[4] Validation-accuracy=0.423377
```

Thanks
Hamid

**Comment by Palang2014** [comment]
@zer0n

**Comment by Palang2014** [comment]
In other words, is there any way to compensate for async SGD in mxnet? Otherwise we should increase the # of epochs to be able to get the same accuracy which as a result mean that we might end up spending the same time on training as we had a single GPU.

**Comment by zer0n** [comment]
Cifar-10 is relatively small. The benefit for training Cifar-10 on multi-GPU is probably none to little.

Async SGD is not the de facto distributed training approach. You have better scalability but lose convergence. The most efficient, i.e. good scaling without compromising much of convergence, distributed algorithm that I've seen is 1-bit SGD implemented in CNTK.

**Comment by piiswrong** [comment]
Single Machine multicard should use sync sgd by default. Have you tested if multiple runs on single card gives the same result?

**Comment by winstywang** [comment]
The problem may be in batch normalization or batch size. batch size per GPU and batch size in total will both affect the speed of convergence.

**Comment by Palang2014** [comment]
@piiswrong  : Almost the same, take a look:

run 1:
```python
2016-01-25 17:44:16,616 Node[0] Start training with [gpu(1)]
2016-01-25 17:44:24,965 Node[0] Epoch[0] Batch [50]	Speed: 824.01 samples/sec	Train-accuracy=0.279219
2016-01-25 17:44:32,874 Node[0] Epoch[0] Batch [100]	Speed: 809.24 samples/sec	Train-accuracy=0.313672
2016-01-25 17:44:40,789 Node[0] Epoch[0] Batch [150]	Speed: 808.64 samples/sec	Train-accuracy=0.330469
2016-01-25 17:44:48,718 Node[0] Epoch[0] Batch [200]	Speed: 807.20 samples/sec	Train-accuracy=0.346406
2016-01-25 17:44:56,662 Node[0] Epoch[0] Batch [250]	Speed: 805.66 samples/sec	Train-accuracy=0.361906
2016-01-25 17:45:04,606 Node[0] Epoch[0] Batch [300]	Speed: 805.63 samples/sec	Train-accuracy=0.372630
2016-01-25 17:45:12,549 Node[0] Epoch[0] Batch [350]	Speed: 805.85 samples/sec	Train-accuracy=0.381228
2016-01-25 17:45:19,073 Node[0] Epoch[0] Train-accuracy=0.388607
2016-01-25 17:45:19,073 Node[0] Epoch[0] Time cost=62.019
2016-01-25 17:45:21,368 Node[0] Epoch[0] Validation-accuracy=0.428797
2016-01-25 17:45:21,389 Node[0] Saved checkpoint to "N/A"
2016-01-25 17:45:29,293 Node[0] Epoch[1] Batch [50]	Speed: 821.29 samples/sec	Train-accuracy=0.457656
2016-01-25 17:45:37,244 Node[0] Epoch[1] Batch [100]	Speed: 804.96 samples/sec	Train-accuracy=0.466562
2016-01-25 17:45:45,183 Node[0] Epoch[1] Batch [150]	Speed: 806.13 samples/sec	Train-accuracy=0.474375
2016-01-25 17:45:53,129 Node[0] Epoch[1] Batch [200]	Speed: 805.46 samples/sec	Train-accuracy=0.479102
2016-01-25 17:46:01,060 Node[0] Epoch[1] Batch [250]	Speed: 806.99 samples/sec	Train-accuracy=0.485781
2016-01-25 17:46:08,989 Node[0] Epoch[1] Batch [300]	Speed: 807.19 samples/sec	Train-accuracy=0.487370
2016-01-25 17:46:16,924 Node[0] Epoch[1] Batch [350]	Speed: 806.53 samples/sec	Train-accuracy=0.492344
2016-01-25 17:46:23,433 Node[0] Epoch[1] Train-accuracy=0.498262
2016-01-25 17:46:23,433 Node[0] Epoch[1] Time cost=62.045
2016-01-25 17:46:25,697 Node[0] Epoch[1] Validation-accuracy=0.529046
```
 
run 2:
```python
2016-01-26 21:49:24,148 Node[0] Start training with [gpu(1)]
2016-01-26 21:49:32,529 Node[0] Epoch[0] Batch [50]     Speed: 821.02 samples/sec       Train-accuracy=0.278281
2016-01-26 21:49:40,477 Node[0] Epoch[0] Batch [100]    Speed: 805.29 samples/sec       Train-accuracy=0.307422
2016-01-26 21:49:48,428 Node[0] Epoch[0] Batch [150]    Speed: 804.96 samples/sec       Train-accuracy=0.328125
2016-01-26 21:49:56,361 Node[0] Epoch[0] Batch [200]    Speed: 806.86 samples/sec       Train-accuracy=0.344141
2016-01-26 21:50:04,312 Node[0] Epoch[0] Batch [250]    Speed: 804.97 samples/sec       Train-accuracy=0.358344
2016-01-26 21:50:12,265 Node[0] Epoch[0] Batch [300]    Speed: 804.75 samples/sec       Train-accuracy=0.369453
2016-01-26 21:50:20,198 Node[0] Epoch[0] Batch [350]    Speed: 806.77 samples/sec       Train-accuracy=0.381004
2016-01-26 21:50:26,716 Node[0] Epoch[0] Train-accuracy=0.389586
2016-01-26 21:50:26,716 Node[0] Epoch[0] Time cost=62.117
2016-01-26 21:50:29,025 Node[0] Epoch[0] Validation-accuracy=0.469937
2016-01-26 21:50:29,045 Node[0] Saved checkpoint to "N/A"
2016-01-26 21:50:36,947 Node[0] Epoch[1] Batch [50]     Speed: 821.51 samples/sec       Train-accuracy=0.462344
2016-01-26 21:50:44,917 Node[0] Epoch[1] Batch [100]    Speed: 803.11 samples/sec       Train-accuracy=0.476484
2016-01-26 21:50:52,864 Node[0] Epoch[1] Batch [150]    Speed: 805.39 samples/sec       Train-accuracy=0.484896
2016-01-26 21:51:00,798 Node[0] Epoch[1] Batch [200]    Speed: 806.61 samples/sec       Train-accuracy=0.486328
2016-01-26 21:51:08,761 Node[0] Epoch[1] Batch [250]    Speed: 803.82 samples/sec       Train-accuracy=0.490563
2016-01-26 21:51:16,721 Node[0] Epoch[1] Batch [300]    Speed: 804.05 samples/sec       Train-accuracy=0.492135
2016-01-26 21:51:24,687 Node[0] Epoch[1] Batch [350]    Speed: 803.39 samples/sec       Train-accuracy=0.496920
2016-01-26 21:51:31,204 Node[0] Epoch[1] Train-accuracy=0.501678
2016-01-26 21:51:31,204 Node[0] Epoch[1] Time cost=62.159
2016-01-26 21:51:33,488 Node[0] Epoch[1] Validation-accuracy=0.530849

```

**Comment by Palang2014** [comment]
@winstywang : Can we set batch size per GPU separately from batch size in MxNet?

**Comment by Palang2014** [comment]
@zer0n : Agreed! Do we have anything similar to 1-bit SGD implemented in MxNet?!

**Comment by piiswrong** [comment]
batch_size_per_gpu = total_batch_size/number_of_gpu

**Comment by mli** [comment]
@Palang2014 1bit sgd has only been tested if there are big fully connected layers. both cntk and amazon echo team used it for speech recognition. but i think it is not necessary for conv networks except for alexnet/vgg.

**Comment by zer0n** [comment]
Very true
